<launch>

<!-- ##################
     #### Live Run ####
     ################## -->
     <!-- NOTE
        Chk USBFS memory (default is 16, not suitable for point grey cam):
          $cat /sys/module/usbcore/parameters/usbfs_memory_mb

        Set USBFS memory
          $sudo sh -c 'echo 1024 > /sys/module/usbcore/parameters/usbfs_memory_mb'

        run point grey cam
          $roslaunch pointgrey_camera_driver camera.launch


     -->

<!--  #################
      ###### BAGS #####
      ################# -->

    <!-- BAG SET 01, captured 29th June, 2017 -->
    <!-- <arg name="bag_path" default="/media/mpkuse/Bulk_Data/other_seqs/pointgrey_color_seq/" /> -->
    <!-- <arg name="bag_file" default="2017-06-29-15-09-10.bag" doc="OK, on stairs for quick testing"/> -->
    <!-- <arg name="bag_file" default="2017-06-29-15-11-31.bag"/> -->
    <!-- <arg name="bag_file" default="2017-06-29-15-16-50.bag" doc="start from UG flr of cyt to lawn and back. not a good seq to test with"/> -->
    <!-- <arg name="bag_file" default="2017-06-29-15-23-43.bag" doc="ok"/> -->
    <!--<arg name="bag_file" default="2017-06-29-15-29-46.bag" doc="traj ok, but not ok to test netvlad"/>-->
    <!-- <arg name="bag_file" default="2017-06-29-15-31-19.bag" doc="dont use"/> -->
    <!-- END OF BAG 01 -->

    <!-- Bag set 02, captured 5th July, 2017 -->
    <!-- <arg name="bag_path" default="/media/mpkuse/Bulk_Data/other_seqs/pointgrey_color_seq/bags02/" /> -->
    <!--<arg name="bag_file" default="2017-07-04-16-28-55.bag" doc="in lift"/>-->
    <!-- <arg name="bag_file" default="2017-07-04-16-32-46.bag" /> -->
    <!-- <arg name="bag_file" default="2017-07-04-16-36-37.bag" doc="bad for testing" /> -->
    <!-- <arg name="bag_file" default="2017-07-04-16-39-32.bag" /> -->
    <!-- <arg name="bag_file" default="2017-07-04-16-41-25.bag" doc="in the BASE. rotated comes to same place" /> -->
    <!-- <arg name="bag_file" default="2017-07-04-16-43-59.bag" doc="outdoor of BASE to towerA and back from the bridge, Only 1 loop missed by my method. Not a good seq for testing"/> -->
    <!-- <arg name="bag_file" default="2017-07-04-16-51-41.bag" /> -->
    <!-- <arg name="bag_file" default="2017-07-04-17-06-23.bag" /> -->
    <!-- END OF BAG 02 -->


    <!-- Bag set 03, Captured on 16th Aug, 2017 -->
    <!-- <arg name="bag_path" default="/media/mpkuse/Bulk_Data/other_seqs/pointgrey_color_seq/bag03/" /> -->
    <!-- <arg name="bag_file" default="2017-08-16-14-54-01.bag" doc="start UG outside CYT, takes stairs down to enterprise center walk back" /> -->
    <!-- <arg name="bag_file" default="2017-08-16-15-00-56.bag" doc="start from SENG towards BASE. Good results!"/> -->
    <!-- <arg name="bag_file" default="2017-08-16-15-06-12.bag" doc="Walking in BASE. Alright! but a little short the seq. 1 wrong"/> -->
    <!-- <arg name="bag_file" default="2017-08-16-15-09-15.bag" doc="!!Perfect seq for testing" /> -->


    <!-- Bag set 05, 4 seq in Tai Po Tsai. Captured on 12th Dec, cloudy weather -->
    <!-- <arg name="bag_path" default="/media/mpkuse/Bulk_Data/other_seqs/pointgrey_color_seq/bag05" /> -->
    <!-- <arg name="bag_file" default="2017-12-12-10-46-42.bag" doc="around 7/11 in tpt. Not good for testing too many silloutes due to pointing camera to the sky"/> -->
    <!-- <arg name="bag_file" default="2017-12-12-10-51-52.bag" doc="useless. No loop" /> -->
    <!-- <arg name="bag_file" default="2017-12-12-10-53-04.bag" doc="building near the garbage dump in tpt" /> -->
    <!-- <arg name="bag_file" default="2017-12-12-11-04-45.bag" doc="!!tpt childrens garden, Another nice seq to test" /> -->
    <!-- <arg name="bag_file" default="2017-12-13-11-12-35.bag" doc="!!indoors of LSK building G/F. Usable" /> -->



    <!-- Bag set 06, Sequences captured with blackbox4, softtry-sequences -->
    <arg name="bag_path" default="/media/mpkuse/Bulk_Data/other_seqs/blackbox4_bags/" />
    <!-- <arg name="bag_file" default="2018-02-19-17-52-36.bag" doc="single loop in lab"/> -->
    <arg name="bag_file" default="2018-03-20-10-40-43.bag" doc="multiple loops in lab"/>
    <!-- <arg name="bag_file" default="2018-04-13-17-59-15.bag" doc="loopy in drone flying area of lab"/> -->
    <!-- <arg name="bag_file" default="2018-04-13-18-28-56.bag" doc=""/> -->
    <!-- <arg name="bag_file" default="2018-04-16-12-03-50.bag" doc="long trajectory in lab"/> -->

    <!-- Bag set 07, Sequences having good circle init, to enable estimating ground plane -->
    <!-- <arg name="bag_path" default="/media/mpkuse/Bulk_Data/other_seqs/blackbox4_bags/" /> -->
    <!-- <arg name="bag_file" default="2018-04-25-11-15-10.bag" doc="ground plane"/> -->
    <!-- <arg name="bag_file" default="2018-04-26-15-03-27.bag" doc="ground plane + long in lab"/> -->
    <!-- <arg name="bag_file" default="2018-04-26-15-28-15.bag" doc="ground plane + long in lab. inside and outside 2ndflr cyt"/> -->


    <!-- Bag set 07. captured with blackbox4 around coffee shop -->
    <!-- <arg name="bag_path" default="/media/mpkuse/Bulk_Data/other_seqs/blackbox4_bags/" /> -->
    <!-- <arg name="bag_file" default="2018-04-15-14-02-59.bag" doc=""/> -->
    <!-- <arg name="bag_file" default="2018-04-15-14-08-06.bag" doc=""/> -->
    <!-- <arg name="bag_file" default="2018-04-15-14-13-17.bag" doc=""/> -->








    <node pkg="rosbag" type="play" name="rosbag" args="$(arg bag_path)/$(arg bag_file)" output="log"/>
    <!--<node name="recorder" pkg="rosbag" type="record" args="-O $(arg bag_path)/cooked_$(arg bag_file) /camera/image_raw /imu/imu /vins_estimator/odometry"/>-->
    <!--<node name="recorder" pkg="rosbag" type="record" args="-O $(arg bag_path)/all_cooked_$(arg bag_file) -a "/>-->

<!-- ##### END BAGS #### -->



<!-- ##################
     ###### VINS ######
     ################## -->
    <arg name="config_path" default = "$(find feature_tracker)/../config/black_box4/blackbox4.yaml" />
    <arg name="vins_path" default = "$(find feature_tracker)/../config/../" />

    <node name="feature_tracker" pkg="feature_tracker" type="feature_tracker" output="log">
        <param name="config_file" type="string" value="$(arg config_path)" />
        <param name="vins_folder" type="string" value="$(arg vins_path)" />
    </node>

    <node name="vins_estimator" pkg="vins_estimator" type="vins_estimator" output="log">
       <param name="config_file" type="string" value="$(arg config_path)" />
       <param name="vins_folder" type="string" value="$(arg vins_path)" />
    </node>

    <!-- 4DOF pose graph -->
    <node name="pose_graph" pkg="pose_graph" type="pose_graph" output="log">
      <param name="config_file" type="string" value="$(arg config_path)" />
      <param name="visualization_shift_x" type="int" value="0" />
      <param name="visualization_shift_y" type="int" value="0" />
      <param name="skip_cnt" type="int" value="0" />
    </node>

    <!-- 6DOF pose graph -->
    <node name="keyframe_pose_graph_slam" pkg="solve_keyframe_pose_graph" type="keyframe_pose_graph_slam" output="log">
    </node>

<!-- ##### END VINS ##### -->




<!--  #############################
      ##### Place Recognition #####
      ############################# -->
      <arg name="bag_color_im_topic" default="/vins_estimator/keyframe_image" /> <!-- Key frames only -->
      <!--<arg name="bag_color_im_topic" default="/camera/image_raw" />--><!-- All input color images -->

      <!--NAP Node -->
      <!-- <node name="nap" pkg="nap" type="nap_graph_modelling.py" output="screen" > -->
      <!-- <node name="nap" pkg="nap" type="nap_daisy_bf.py" output="screen" > -->
      <!-- <node name="nap" pkg="nap" type="nap_robustdaisy_bf.py" output="screen" >
        <remap to="$(arg bag_color_im_topic)" from="/semi_keyframes"/>
        <param name="config_file" type="string" value="$(arg config_path)" />
      </node> -->

      <!-- Multiprocess and minimal nap node -->
      <!-- <node name="nap" pkg="nap" type="nap_multiproc.py" output="screen" > -->
      <node name="nap" pkg="nap" type="nap_multiproc2.py" output="log" >
        <remap to="$(arg bag_color_im_topic)" from="/semi_keyframes"/>
        <param name="config_file" type="string" value="$(arg config_path)" />
        <param name="debug_output_dir" type="string" value="/home/mpkuse/Desktop/bundle_adj" />
      </node>


      <!-- DBOW node by mpkuse -->
       <!-- <node name="dbow_node" pkg="nap" type="dbow3_naive" output="screen">
        <remap to="$(arg bag_color_im_topic)" from="/color_image_inp" />
        <remap from="/colocation_dbow" to="/colocation" />
      </node> -->

      <!-- DBOW node by Qin Tong -->
      <!-- TODO: In tony's node add code to writes images to file and also writeout all loop_candidates. But essentially this is same as my DBOW node, difference is in the way comparison thresh is set -->
      <!-- <node name="dbow_qintony_node" pkg="loop_detector" type="loop_detector" output="screen"> -->
        <!-- <remap from="/raw_graph_edge" to="/colocation" /> -->
    <!-- </node> -->



<!-- ##### END Place Recognition ##### -->



<!-- ###########################
     ##### Geometry Node   #####
     ###########################
            subscribes to `vins/camera_pose`, `3d_point_cloud`, `napMsg`  -->
    <node name="pgraph_opt" pkg="nap" type="pose_graph_opt_node" output="screen">
        <param name="config_file" type="string" value="$(arg config_path)" />
        <!-- Commentout next line to remove the places-feedback-->
        <!-- <remap from="/colocation_chatter" to="/colocation" /> -->
    </node>

<!-- ##### END Pose Graph Optimization ##### -->



<!--  ################
      ##### RVIZ #####
      ################ -->

    <!-- Qin Tong's Visualization -->
    <!--<node name="rvizvisualisation" pkg="rviz" type="rviz" output="log" args="-d $(find vins_estimator)/../config/vins_rviz_config.rviz" />-->


    <!-- <node name="rvizvisualisation" pkg="rviz" type="rviz" output="log" args="-d $(find nap)/rviz/my_pose_graph.rviz" /> -->

    <!-- Display/Publish the colocation image. Subscribes to /raw_graph_visual_edge -->
    <!-- <node name="colocation_viz_lite" pkg="nap" type="colocation_viz_lite.py" output="log" /> -->


    <!-- Old AR (node) has static markers -->
    <!-- <node pkg="ar_demo" type="ar_demo_node" name="ar_demo_node" output="log">
      <remap from="~image_raw" to="/pg_17302081/image" doc="input image"/>
      <remap from="~camera_path" to="/pose_graph/pose_graph_camera_path"/>

      <param name="calib_file" type="string" value="$(arg config_path)"/>

      <param name="box_length" type="double" value="0.8"/>
      <param name="in_x" type="double" value="-8.5"/>
      <param name="in_y" type="double" value="-.8"/>
      <param name="in_z" type="double" value="-1.2"/>


      <param name="use_undistored_img" type="bool" value="false"/>
  </node>
  -->


<!-- ##################################
     ##### Augmented Reality (AR) #####
     ################################## -->

    <!-- AR Node with interactive markers. Multiple obj can be specified as ';' separated value -->
    <arg name="ar_obj_list" default="cube.obj" />
    <arg name="ar_mesh_scaling_list" default=".5" />
    <arg name="ar_controls_scaling_list" default="1.5" />
    <arg name="ar_mesh_initial_positions_list" default="-8,-1,-0.5" />

    <arg name="ar_input_camera_image" default="/pg_17302081/image" doc="video feed"/>
    <arg name="camera_navmsg_path" default="/pose_graph/pose_graph_camera_path" doc="4DOF. If using this make sure pose_graph node is runnning" />
    <!-- <arg name="camera_navmsg_path" default="/keyframe_pose_graph_slam/opt_path" doc="6DOF. If using this make sure solve_posegraph_opt node is running" /> -->


    <node pkg="ar_demo" type="ar_demo_node2" name="ar_demo_node" output="log">
      <remap from="~image_raw" to="$(arg ar_input_camera_image)" doc="input image"/>
      <remap from="~camera_path" to="$(arg camera_navmsg_path)"/>
      <remap from="~object_mesh_pose" to="/interactive_marker_server/object_mesh_pose"/>

      <param name="calib_file" type="string" value="$(arg config_path)"/>

      <param name="obj_list" type="string" value="$(arg ar_obj_list)"/>
      <param name="mesh_scaling_list" type="string" value="$(arg ar_mesh_scaling_list)"/>
      <param name="mesh_initial_positions_list" type="string" value="$(arg ar_mesh_initial_positions_list)" />

  </node>

  <node pkg="ar_demo" type="interactive_marker_server" name="interactive_marker_server" output="log" >
      <param name="obj_list" type="string" value="$(arg ar_obj_list)"/>
      <param name="mesh_scaling_list" type="string" value="$(arg ar_mesh_scaling_list)"/>
      <param name="controls_scaling_list" type="string" value="$(arg ar_controls_scaling_list)"/>
      <param name="mesh_initial_positions_list" type="string" value="$(arg ar_mesh_initial_positions_list)" />
  </node>


  <!-- Subscribes to keyframe pointcloud (from vins_estimator node) and publishes marker message to show 1st 50 frames point cloud and ground plane -->
  <node name="plane_detect" pkg="plane_detect" type="plane_detect_node.py"  output="log">
    <param name="RANSAC_n_iterations" type="int" value="2000"/>
    <param name="RANSAC_sigma" type="double" value="0.02"/>
    <param name="RANSAC_q" type="int" value="100"/>
    <param name="WAIT_N_FRAMES" type="int" value="65"/>

    <param name="PLANE_r_color" type="double" value="0.0"/>
    <param name="PLANE_g_color" type="double" value="1.0"/>
    <param name="PLANE_b_color" type="double" value="1.0"/>
    <param name="PLANE_a_color" type="double" value=".8"/>
    <param name="PLANE_xspan" type="double" value="5"/>
    <param name="PLANE_yspan" type="double" value="5"/>
  </node>




<!-- ##### End Visualization ##### -->





</launch>
